numpyro/diagnostics.py

```diff
@@ -0,0 +1,217 @@
+    :returns numpy.ndarray: split R-hat of ``x``.
+    """
+    assert x.ndim >= 2
+    assert x.shape[1] >= 4
+
+    N_half = x.shape[1] // 2
+    new_input = onp.concatenate([x[:, :N_half], x[:, -N_half:]], axis=0)
+    split_rhat = gelman_rubin(new_input)
+    return split_rhat
+
+
+def _fft_next_fast_len(target):
+    # find the smallest number >= N such that the only divisors are 2, 3, 5
+    # works just like scipy.fftpack.next_fast_len
+    if target <= 2:
+        return target
+    while True:
+        m = target
+        while m % 2 == 0:
+            m //= 2
+        while m % 3 == 0:
+            m //= 3
+        while m % 5 == 0:
+            m //= 5
+        if m == 1:
+            return target
+        target += 1
+
+
+def autocorrelation(x, axis=0):
+    """
+    Computes the autocorrelation of samples at dimension ``axis``.
+
+    :param torch.Tensor input: the input tensor.
+    :param int dim: the dimension to calculate autocorrelation.
+    :returns torch.Tensor: autocorrelation of ``input``.
+    """
+    # Ref: https://en.wikipedia.org/wiki/Autocorrelation#Efficient_computation
+    # Adapted from Stan implementation
+    # https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/autocorrelation.hpp
+    N = x.shape[axis]
+    M = _fft_next_fast_len(N)
+    M2 = 2 * M
+
+    # transpose axis with -1 for Fourier transform
+    x = onp.swapaxes(x, axis, -1)
+
+    # centering x
+    centered_signal = x - x.mean(axis=-1, keepdims=True)
+
+    # Fourier transform
+    freqvec = onp.fft.rfft(centered_signal, n=M2, axis=-1)
+    # take square of magnitude of freqvec (or freqvec x freqvec*)
+    freqvec_gram = freqvec * onp.conjugate(freqvec)
+    # inverse Fourier transform
+    autocorr = onp.fft.irfft(freqvec_gram, n=M2, axis=-1)
+
+    # truncate and normalize the result, then transpose back to original shape
+    autocorr = autocorr[..., :N]
+    autocorr = autocorr / onp.arange(N, 0., -1)
+    autocorr = autocorr / autocorr[..., :1]
+    return onp.swapaxes(autocorr, axis, -1)
+
+
+def autocovariance(x, axis=0):
+    """
+    Computes the autocovariance of samples at dimension ``axis``.
+
+    :param numpy.ndarray x: the input array.
+    :param int axis: the dimension to calculate autocorrelation.
+    :returns numpy.ndarray: autocorrelation of ``x``.
+    """
+    return autocorrelation(x, axis) * x.var(axis=axis, keepdims=True)
+
+
+def effective_sample_size(x):
+    """
+    Computes effective sample size of input ``x``, where the first dimension of
+    ``x`` is chain dimension and the second dimension of ``x`` is draw dimension.
+
+    **References:**
+    [1] `Introduction to Markov Chain Monte Carlo`,
+        Charles J. Geyer
+    [2] `Stan Reference Manual version 2.18`,
+        Stan Development Team
+
+    :param numpy.ndarray x: the input array.
+    :returns numpy.ndarray: effective sample size of ``x``.
+    """
+    assert x.ndim >= 2
+    assert x.shape[1] >= 2
```

Comment:
I think we are assuming here that `x.shape[1]` corresponds to the draw dimension which might not be correct with multiple chains?

Reply:
Yes, first dimension is chain dimension. It can be that `x.shape[0] == 1` (single chain).

Reply:
Hmm..sorry my earlier comment was wrong. I meant that the draw or batch dimension is either the first dim (0) or the second (1) with multiple chains. Just want to make sure that this will aggregate across all the chains when a chain dimension is present.

Reply:
Yup, split_gelman_rubin and effective_sample_size are chain diagnostics (but can work with num_chains=1), so the result is aggregated across all chains. In this PR, only `summary` does not support `chain` dimension yet.